vm/src/obj/objstr.rs

```diff
@@ -786,25 +786,33 @@ impl PyString {
 
     #[pymethod]
     fn splitlines(&self, args: pystr::SplitLinesArgs, vm: &VirtualMachine) -> PyObjectRef {
-        let mut elements = vec![];
-        let mut curr = "".to_owned();
-        let mut chars = self.value.chars().peekable();
-        while let Some(ch) = chars.next() {
-            if ch == '\n' || ch == '\r' {
-                if args.keepends {
-                    curr.push(ch);
+        let keep = if args.keepends { 1 } else { 0 };
```

Comment:
It seems like you have the same code duplicated. Maybe share it?

Reply:
I am struggling with how to share logics around str::chars() and [u8]::iter()

Do you have any idea to do it without macro?

Reply:
Maybe define a generic `get_iterator` in `PyCommonString`?

Reply:
In that case the type will be `dyn std::iter::Iterator` and i am afraid it would have additional runtime overhead

Reply:
You could have an associated type on `PyCommonString` for the inner character type, either `u8` or `char`, and then another associated type with a bound `Iterator<Item = Self::Character>`.

Reply:
I tried this several days ago but It was not easy due to lifetime of `Chars`

Reply:
You could just have a function in `pystr.rs` that takes a `Iterator` and does the operation, it doesn't need to be a method. Or just live with the code duplication for now, it's only 2 instances.